### PR TITLE
⛩️ MLflow Gateway ⛩️ 

### PR DESCRIPTION
**Pull request by harupy**
# When we merging commits on this branch, please make sure to use "rebase and merge" or "create a merge commit" after rebasing this branch on master.

### CODE REVIEW

File: mlflow/gateway/config.py [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+            "The api key provided is not a string. Please provide either an environment "
+            "variable key, a path to a file containing the api key, or the api key itself"
+        )
+
+    # try reading as an environment variable
+    env_var_attempt = api_key_input[1:] if api_key_input.startswith("$") else api_key_input
+
+    env_var = os.getenv(env_var_attempt)
+    if env_var:
+        return env_var
```
Comment (harupy) [inline mlflow/gateway/config.py:98]: I want to make sure the logig here is correct. In the current logic, `openai_api_key: OPENAI_API_KEY` means `use the env var OPENAI_API_KEY`.

File: mlflow/gateway/runner.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+                "--bind",
+                f"{self.host}:{self.port}",
+                "--workers",
+                str(self.workers),
+                "--worker-class",
+                "uvicorn.workers.UvicornWorker",
+                f"{app.__name__}:create_app_from_env()",
+                "--reload",
+                "--log-level",
+                "debug",
```
Comment (harupy) [inline mlflow/gateway/runner.py:73]: This needs to be configurable.

File: mlflow/gateway/config.py [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+    Provider.ANTHROPIC: AnthropicConfig,
+    Provider.DATABRICKS_SERVING_ENDPOINT: DatabricksConfig,
+    Provider.MLFLOW: MLflowConfig,
+    Provider.CUSTOM: CustomConfig,
+}
+
+
+class ModelInfo(BaseModel, extra=Extra.forbid):
+    name: Optional[str] = None
+    provider: Provider = Provider.CUSTOM
```
Comment (harupy) [inline mlflow/gateway/config.py:73]: Should the default provider be `custom`?

File: mlflow/gateway/constants.py [hunk 4]
```diff
@@ -0,0 +1 @@
+PROVIDERS = {"openai", "anthropic", "databricks_serving_endpoint", "mlflow_served_model"}
```
Comment (harupy) [inline mlflow/gateway/constants.py:1]: This should be dinaymically created be from the `Provider` model.

**Comment by harupy** [comment]
Let's rebase this branch on master periodically to minimize conflicts.

**Comment by harupy** [comment]
Rebased on master

File: mlflow/gateway/config.py [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+    OPENAI = "openai"
+    ANTHROPIC = "anthropic"
+    DATABRICKS_SERVING_ENDPOINT = "databricks_serving_endpoint"
+    MLFLOW = "mlflow"
+
+
+class RouteType(str, Enum):
+    CUSTOM = "custom"
+    LLM_V1_COMPLETIONS = "llm/v1/completions"
+    LLM_V1_CHAT = "llm/v1/chat"
```
Comment (harupy) [inline mlflow/gateway/config.py:31]: @dbczumar @BenWilson2 Are we missing `embeddings` here?

Comment (dbczumar) [reply mlflow/gateway/config.py:31]: Yes, we should adding embeddings. Embeddings are P0, chat is P1

File: mlflow/gateway/config.py (hunk 3 above)
Comment (dbczumar) [reply mlflow/gateway/config.py:73]: No, I think we can remove it

File: mlflow/gateway/config.py (hunk 1 above)
Comment (dbczumar) [reply mlflow/gateway/config.py:98]: Yes, I agree; we should only use an env var when the key starts with a dollar sign

File: mlflow/gateway/constants.py (hunk 4 above)
Comment (dbczumar) [reply mlflow/gateway/constants.py:1]: +1

File: mlflow/gateway/runner.py (hunk 2 above)
Comment (dbczumar) [reply mlflow/gateway/runner.py:73]: Agreed

File: mlflow/gateway/config.py (hunk 3 above)
Comment (harupy) [reply mlflow/gateway/config.py:73]: https://github.com/mlflow/mlflow/pull/8709

File: mlflow/gateway/runner.py (hunk 2 above)
Comment (harupy) [reply mlflow/gateway/runner.py:73]: https://github.com/mlflow/mlflow/pull/8701

File: mlflow/gateway/constants.py (hunk 4 above)
Comment (BenWilson2) [reply mlflow/gateway/constants.py:1]: yeah it was a placeholder :)

File: mlflow/gateway/config.py (hunk 5 above)
Comment (harupy) [reply mlflow/gateway/config.py:31]: Filed https://github.com/mlflow/mlflow/pull/8711

File: mlflow/gateway/constants.py (hunk 4 above)
Comment (harupy) [reply mlflow/gateway/constants.py:1]: Filed #8712

File: mlflow/gateway/config.py (hunk 1 above)
Comment (harupy) [reply mlflow/gateway/config.py:98]: Filed #8710

**Comment by harupy** [comment]
Before merge, let's rebase this branch on master to make sure Gateway doesn't break anything.

**Pull request merged**